File: AI_Codes/queen.cpp
#include <bits/stdc++.h>
using namespace std;

#define N 8

void printSolution(int board[N][N]) {
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++)
            cout << board[i][j] << " ";
        cout << endl;
    }
}

bool isSafe(int row, int col, int board[N][N]) {
    for (int i = 0; i < col; i++)
        if (board[row][i]) return false;

    for (int i=row, j=col; i>=0 && j>=0; i--, j--)
        if (board[i][j]) return false;

    for (int i=row, j=col; j>=0 && i<N; i++, j--)
        if (board[i][j]) return false;

    return true;
}

bool solve(int board[N][N], int col) {
    if (col >= N)
        return true;

    for (int i = 0; i < N; i++) {
        if (isSafe(i, col, board)) {
            board[i][col] = 1;

            if (solve(board, col + 1))
                return true;

            board[i][col] = 0;
        }
    }
    return false;
}

int main() {
    int board[N][N] = {0};

    if (!solve(board, 0))
        cout << "No solution";
    else
        printSolution(board);

    return 0;
}

File: AI_Codes/queen.py
# Check if placing a queen is safe
def is_safe(board, row, col):
    """Check if placing a queen at (row, col) is safe."""
    for i in range(row):
        # Same column OR same diagonal
        if (board[i] == col or
            board[i] - i == col - row or
            board[i] + i == col + row):
            return False
    return True


# Backtracking function
def solve_n_queens(board, row, n):
    """Recursively place queens row by row using backtracking."""
    
    # Base case: all queens placed
    if row == n:
        return True

    # Try each column
    for col in range(n):
        if is_safe(board, row, col):
            board[row] = col  # Place queen

            if solve_n_queens(board, row + 1, n):
                return True

            board[row] = -1  # Backtrack

    return False


# Main function
def n_queens(n):
    """Initialize the board and solve the N-Queens problem."""
    
    board = [-1] * n  # -1 means no queen placed

    if solve_n_queens(board, 0, n):
        print("Solution found:")
        print(board)
    else:
        print("No solution exists")


# Run for n = 8
n = 8
n_queens(n)